Git-Closure/English.md

```diff
@@ -0,0 +1,168 @@
+
+![Git Loop](Git-Loop.svg)
+
+<!--
+```graphviz
+digraph {
+    node [shape=circle];
+    rankdir="LR";
+
+    1 -> 1
+}
+```
+
+-->
+
+You can close the branch over any commit - Git is smart, it calculates the
+differences and merges everything correctly:
+
+![Git Closure Custom](Git-Closure-Custom.svg)
+
+<!--
+```graphviz
+digraph {
+    node [shape=circle];
+    rankdir="LR";
+
+    1 -> 2 -> 3 -> 6;
+    1 -> 4 -> 5 -> 6;
+}
+```
+
+-->
+
+## How to use?
+
+The `merge` command does not include the functionality of closures, since for
+the first case the branch will be
+[fast-forwarded](https://stackoverflow.com/q/9069061/1046374), and for the
+second case nothing will be done (`git already up to date`).
+
+In order not to change the old behavior, the developers decided to introduce a
+closure command:
+
+```
+git closure -s $source_commit -d $dest_commit -m $message
+```
+
+The first argument `-s $source_commit` sets the hash of the commit from which
+you want to stretch the loop, and the second (optional) `-d $dest_commit` sets
+the commit into which the loop will be closed. If it is absent, the closure
+occurs in the current check-out branch. The `-m $message` argument sets a
+closure message, like `failed feature, revert to origin`. However, the
+`--allow-empty-message` option is also available, which allows commits without
+messages. By default, Git allows only one closure for a pair of commits. To
+bypass this limitation, the `--allow-multiple-closures` option is available.
+
+![Git Multiple Closure](Git-Closure-Multiple.svg)
+
+After the command is executed, Git calculates the changes, and in the final
+commit a double diff will be shown: from the base and closing branches. In the
+general case, it is an n-dimensional diff, that is, there can be as many
+closures as you wish. closure-commit is similar to merge-commit with the only
+difference that it contains several messages, not one.
+
+Unfortunately, existing Git GUI's do not have good support of closures.
+[GitExtensions](https://github.com/gitextensions/gitextensions) preview version
+displays merge curves instead of elegant arcs. Take a look on the new fields such
+as `Closure message` and `Closure diff`:
+
+![Git Closure Merge](Git-Closure-Merge.png)
+
+It is worth noting that the `closure` command always changes history (as now Git
+is a full-fledged time machine!), so now it is possible to push branches only
+with the `--force` option, or the safe `--force-with-lease` option.
+
+Rebase is also available for looped branches, although the logic for
+recalculating commits is complicated.
+
+Also the `auto` option allows automatic closures of all the old branches. In
+this case, the closing commit is the one starting the branch. With Git IDE
+plugins, closures can be run periodically. In
+GitExtensions there is a similar plugin **Delete obsolete branches**.
+
+## What IT companies think
+
+Large IT companies: Google, Facebook, Apple, DeepMind, Positive Technologies,
```

Comment:
I like Positive Technologies in this list, though I strongly suspect it's dwarfed by these peers in terms of market capitalization. :smile:

Reply:
Sure, but I think it's ok for April 1 Fool's Day article :smile: